src/formats/XYZ.cpp

```diff
@@ -252,6 +304,115 @@ std::string write_extended_comment_line(const Frame& frame) {
     return result;
 }
 
+static bool is_valid_property_name(const std::string& name) {
+    if (name.empty()) {
+        return false;
+    }
+
+    if (!is_ascii_letter(name[0])) {
+        return false;
+    }
+
+    for (auto c: name) {
+        if (!(is_ascii_alphanumeric(c) || c == '_')) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+properties_list_t get_atom_properties(const Frame& frame) {
+    if (frame.size() == 0) {
+        return {};
+    }
+
+    auto all_properties = std::map<std::string, Property::Kind>();
+    auto partially_defined_already_warned = std::set<std::string>();
+
+    const auto& first_atom = frame[0];
+    for (const auto& property: first_atom.properties()) {
+        if (!is_valid_property_name(property.first)) {
+            warning(
+                "Extended XYZ", "'{}' is not a valid property name for extended "
+                "XYZ, is will not be saved",
+                property.first, 0
+            );
+            partially_defined_already_warned.insert(property.first);
+            continue;
+        }
+
+        if (property.second.kind() == Property::STRING) {
+            if (should_be_quoted(property.second.as_string())) {
+                warning(
+                    "Extended XYZ", "string value for property '{}' on atom {} "
+                    "can not be be saved as an atomic property",
+                    property.first, 0
+                );
+                continue;
+            }
+        }
+
+        all_properties.emplace(property.first, property.second.kind());
+    }
+
+    for (size_t i=1; i<frame.size(); i++) {
+        const auto& atom = frame[i];
+
+        // warn for properties defined on this atom but not on others
+        for (const auto& property: atom.properties()) {
+            if (all_properties.count(property.first) == 0) {
+                if (partially_defined_already_warned.count(property.first) == 0) {
+                    warning(
+                        "Extended XYZ",
+                        "property '{}' is only defined for a subset of atoms, it will not be saved",
+                        property.first
+                    );
+                    partially_defined_already_warned.insert(property.first);
+                }
+            }
+        }
```

Comment:
Might be a little too pedantic but I feel there is a lot of iterating going on.
Could this be moved after the next loop and only be done if `atom.properties().size() > all_properties` or if at least one property was present in all previous atoms but not this one?
So in most sane cases this would simply be skipped. Not sure if this has any real-world performance impacts.

Reply:
I'll try moving it, although I don't think this will have performance impact for any real world scenario. You would need to have a lot of atoms in the frame, all with very different properties.